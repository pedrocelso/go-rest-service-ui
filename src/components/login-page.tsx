import React, { Component } from 'react';
import jwt from 'jsonwebtoken'
import './login-page.scss'

interface LoginPageState {
  name?: string;
  email?: string;
}

class LoginPage extends Component<{}, LoginPageState> {
  constructor(props: any) {
    super(props)
    this.state = {email: ``, name: ``}
  }

  handleEmailChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({email: e.currentTarget.value})
  }

  handleNameChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({name: e.currentTarget.value})
  }

  generateToken = (): string => {
    const {state} = this
    const jwtSecret = process.env.REACT_APP_JWT_SECRET as string

    const jwtToken = jwt.sign({email: state.email, name: state.name}, jwtSecret)
    return jwtToken;
  }

  redirect = () => {
    const token = this.generateToken();
    sessionStorage.setItem(`jwtToken`, token);
    document.location.pathname = `/`
  }

  render() {
    return (
      <div id="myModal" className="">
        <div className="modal-dialog modal-login">
          <div className="modal-content">
            <div className="modal-header">				
              <h4 className="modal-title">Sign In</h4>
              <button type="button" className="close" data-dismiss="modal" aria-hidden="true">&times;</button>
            </div>
            <div className="modal-body">
                <div className="form-group">
                  <div className="input-group">
                    <span className="input-group-addon"><i className="fa fa-user"></i></span>
                    <input type="text" className="form-control" name="name" placeholder="Name" value={this.state.name} onChange={this.handleNameChange} />
                  </div>
                </div>
                <div className="form-group">
                  <div className="input-group">
                    <span className="input-group-addon"><i className="fa fa-user"></i></span>
                    <input type="text" className="form-control" name="email" placeholder="E-mail" value={this.state.email} onChange={this.handleEmailChange} />
                  </div>
                </div>
                {/*<div className="form-group">
                  <div className="input-group">
                    <span className="input-group-addon"><i className="fa fa-lock"></i></span>
                    <input type="text" className="form-control" name="password" placeholder="Password" />
                  </div>
                </div>*/}
                <div className="form-group">
                  <button type="button" className="btn btn-primary btn-block btn-lg" onClick={this.generateToken}>Sign In</button>
                </div>
            </div>
          </div>
        </div>
      </div>   
    )
  }
}

export default LoginPage;
