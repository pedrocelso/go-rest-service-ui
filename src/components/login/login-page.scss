.modal__dialog {
  width: 30rem;
  margin: auto;
}

.modal__grid {
  margin: 5rem;
  h4 {
    padding: 2rem;
  }
}

.paper-content {
  width: fit-content;
  margin: 1.75rem auto !important;
}

.btnLogin {
  padding: .5rem 1rem !important;
  width: 100%;
}

input {
  border-radius: 0 !important;
  border-top: 0 !important;
  border-left: 0 !important;
  border-right: 0 !important;
}