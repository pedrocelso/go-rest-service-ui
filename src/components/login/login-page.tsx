import React, { Component } from 'react';
import { connect } from 'react-redux'
import { toast } from 'react-toastify';
import './login-page.scss'
import { authenticate } from './login-actions'
import {UserService, User} from '../../services/user'
import pathOr from 'ramda/es/pathOr';

interface LoginPageProps {
  authenticate: typeof authenticate
  redirect: () => void
  service: UserService;
}

interface LoginPageState {
  email: string
  password: string
}

const emailRegexp = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
const isValidEmail = (email: string) => emailRegexp.test(email)

export class LoginPage extends Component<LoginPageProps, LoginPageState> {
  handleEmailChange = (e: React.FormEvent<HTMLInputElement>) => {
    const {value} = e.currentTarget
    this.setState({email: value})
  }

  handlePasswordChange = (e: React.FormEvent<HTMLInputElement>) => {
    const {value} = e.currentTarget
    this.setState({password: value})
  }

  redirect = () => {
    const {authenticate, redirect, service} = this.props
    const {email, password} = this.state
    if (isValidEmail(email) && !!password) {

      service.authenticate({email, password})
      .fork(
        (e) => {
          if (e.statusCode === 401) {
            this.notify(`Wrong email/password!`, -1)
          }
        },
        ({token}) => {
          sessionStorage.setItem(`jwtToken`, token)
          authenticate(token)
          redirect()
        }
      )
    } else {
      this.notify(`Invalid credentials!`, -1)
    }
  }

  notify = (msg: string, feeling: number) => {
    switch(feeling) {
      case(-1):
        toast.error(msg);
    }
  }

  render() {
    const {email} = pathOr(``, [`state`], this)

    return (
      <div id="myModal" className="">
        <div className="modal-dialog modal-login">
          <div className="modal-content">
            <div className="modal-header">				
              <h4 className="modal-title">Sign In</h4>
            </div>
            <div className="modal-body">
              <div className="form-group">
                <div className="input-group">
                  <span className="input-group-addon"><i className="fa fa-user"></i></span>
                  <input type="text" className={`form-control ${!isValidEmail(email) ? "is-invalid" : ""}`} name="email" placeholder="E-mail" onChange={this.handleEmailChange} />
                </div>
              </div>
              <div className="form-group">
                <div className="input-group">
                  <span className="input-group-addon"><i className="fa fa-lock"></i></span>
                  <input type="password" className="form-control" name="password" placeholder="Password" onChange={this.handlePasswordChange} />
                </div>
              </div>
              <div className="form-group">
                <button type="button" className="btn btn-primary btn-block btn-lg" onClick={this.redirect}>Sign In</button>
              </div>
            </div>
          </div>
        </div>
      </div>   
    )
  }
}

const mapDispatchToProps = { authenticate }
export default connect(null, mapDispatchToProps)(LoginPage)