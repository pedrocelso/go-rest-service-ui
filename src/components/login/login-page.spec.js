import React from 'react';
import { shallow } from 'enzyme'
import sinon from 'sinon'

import { LoginPage } from './login-page'

const baseProps = {
  authenticate: sinon.fake(),
  service: {
    authenticate: (_) => ({
      fork: (rej, res) => {
        res({ token: `token!` })
      }
    })
  }
}

describe(`<LoginPage />`, () => {
  it(`Should render without errors`, () => {
    const component = shallow(
      <LoginPage {...baseProps} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`.handlePasswordChange() - Setting valid name`, () => {
    const props = {
      ...baseProps
    }
    const wrapper = shallow(<LoginPage {...props} />);

    wrapper.instance().handlePasswordChange({ currentTarget: { value: `bororoska` } })
    expect(wrapper.state().password).toEqual(`bororoska`)
  });

  it(`.handleEmailChange() - Setting valid email`, () => {
    const props = {
      ...baseProps
    }
    const wrapper = shallow(<LoginPage {...props} />);
    wrapper.instance().handleEmailChange({ currentTarget: { value: `bororoska@test.com` } })
    expect(wrapper.state().email).toEqual(`bororoska@test.com`)
  });

  it(`.login() - valid state`, () => {
    const authenticate = sinon.spy()

    const props = {
      ...baseProps,
      authenticate
    }

    const wrapper = shallow(<LoginPage {...props} />);

    wrapper.setState({ password: 'bar', email: 'foo@bar.com' });

    wrapper.instance().login();
    expect(authenticate.called).toBeTruthy();
  });

  it(`.login() - invalid email`, () => {
    const props = {
      ...baseProps,
      authenticate: sinon.spy()
    }

    const wrapper = shallow(<LoginPage {...props} />);
    wrapper.setState({ password: 'bar', email: 'foo@bar' });

    wrapper.instance().login();
    expect(props.authenticate.called).toBeFalsy();
  });

  it(`.login() - empty email`, () => {
    const props = {
      ...baseProps,
      authenticate: sinon.spy()
    }

    const wrapper = shallow(<LoginPage {...props} />);
    wrapper.setState({ password: 'bar' });

    wrapper.instance().login();
    expect(props.authenticate.called).toBeFalsy();
  });

  it(`.login() - empty name`, () => {
    const props = {
      ...baseProps,
      authenticate: sinon.spy()
    }

    const wrapper = shallow(<LoginPage {...props} />);
    wrapper.setState({ email: 'foo@bar.com' });

    wrapper.instance().login();
    expect(props.authenticate.called).toBeFalsy();
  });

  it(`.login() - bad response from server`, () => {
    const authenticate = sinon.spy()
    const service = {
      authenticate: (_) => ({
        fork: (rej, res) => {
          rej({ statusCode: 401 })
        }
      })
    }

    const props = {
      ...baseProps,
      service,
      authenticate
    }

    const wrapper = shallow(<LoginPage {...props} />);

    wrapper.setState({ password: 'bar', email: 'foo@bar.com' });

    wrapper.instance().login();
    expect(wrapper.state().loading).toBeFalsy()
    expect(authenticate.called).toBeFalsy();
  });
});