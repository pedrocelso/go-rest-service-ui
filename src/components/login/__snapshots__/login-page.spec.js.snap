// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> Should render without errors 1`] = `
<div
  className=""
  id="myModal"
>
  <div
    className="modal-dialog modal-login"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Sign In
        </h4>
      </div>
      <div
        className="modal-body"
      >
        <div
          className="form-group"
        >
          <div
            className="input-group"
          >
            <span
              className="input-group-addon"
            >
              <i
                className="fa fa-user"
              />
            </span>
            <input
              className="form-control "
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="form-group"
        >
          <div
            className="input-group"
          >
            <span
              className="input-group-addon"
            >
              <i
                className="fa fa-user"
              />
            </span>
            <input
              className="form-control "
              name="email"
              onChange={[Function]}
              placeholder="E-mail"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="form-group"
        >
          <button
            className="btn btn-primary btn-block btn-lg"
            onClick={[Function]}
            type="button"
          >
            Sign In
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
